modules/git/repo_tag.go

```diff
@@ -161,37 +162,83 @@ func (repo *Repository) GetTag(name string) (*Tag, error) {
 
 // GetTagInfos returns all tag infos of the repository.
 func (repo *Repository) GetTagInfos(page, pageSize int) ([]*Tag, int, error) {
-	// TODO this a slow implementation, makes one git command per tag
-	stdout, err := NewCommand(repo.Ctx, "tag").RunInDir(repo.Path)
+	forEachRefFmt := foreachref.NewFormat("objecttype", "refname:short", "object", "objectname", "creator", "contents", "contents:signature")
+
+	stdout, err := NewCommand(repo.Ctx, "for-each-ref", "--format", forEachRefFmt.Flag(), "--sort", "-*creatordate", "refs/tags").RunInDir(repo.Path)
 	if err != nil {
 		return nil, 0, err
 	}
 
-	tagNames := strings.Split(strings.TrimRight(stdout, "\n"), "\n")
-	tagsTotal := len(tagNames)
-
-	if page != 0 {
-		tagNames = util.PaginateSlice(tagNames, page, pageSize).([]string)
-	}
-
-	tags := make([]*Tag, 0, len(tagNames))
-	for _, tagName := range tagNames {
-		tagName = strings.TrimSpace(tagName)
-		if len(tagName) == 0 {
-			continue
+	var tags []*Tag
+	parser := forEachRefFmt.Parser(strings.NewReader(stdout))
+	for {
+		ref := parser.Next()
+		if ref == nil {
+			break
 		}
 
-		tag, err := repo.GetTag(tagName)
+		tag, err := parseTagRef(ref)
 		if err != nil {
-			return nil, tagsTotal, err
+			return nil, 0, fmt.Errorf("GetTagInfos: parse tag: %w", err)
 		}
-		tag.Name = tagName
 		tags = append(tags, tag)
 	}
+	if err := parser.Err(); err != nil {
+		return nil, 0, fmt.Errorf("GetTagInfos: parse output: %w", err)
+	}
+
+	tagsTotal := len(tags)
+	if page != 0 {
+		tags = util.PaginateSlice(tags, page, pageSize).([]*Tag)
+	}
+	// TODO shouldn't be necessary since we're running with '--sort'?
 	sortTagsByTime(tags)
 	return tags, tagsTotal, nil
 }
 
+func parseTagRef(ref map[string]string) (tag *Tag, err error) {
+	tag = &Tag{
+		Type: ref["objecttype"],
+		Name: ref["refname:short"],
+	}
+
+	tag.ID, err = NewIDFromString(ref["objectname"])
+	if err != nil {
+		return nil, fmt.Errorf("parse objectname '%s': %v", ref["objectname"], err)
+	}
+
+	if tag.Type == "commit" {
+		// lightweight tag
+		tag.Object = tag.ID
+	} else {
+		// annotated tag
+		tag.Object, err = NewIDFromString(ref["object"])
+		if err != nil {
+			return nil, fmt.Errorf("parse object '%s': %v", ref["object"], err)
+		}
+	}
+
+	tag.Tagger, err = newSignatureFromCommitline([]byte(ref["creator"]))
+	if err != nil {
+		return nil, fmt.Errorf("parse tagger: %w", err)
+	}
+
+	tag.Message = ref["contents"]
+	// strip PGP signature if present in contents field
+	pgpStart := strings.Index(tag.Message, beginpgp)
+	if pgpStart >= 0 {
+		tag.Message = tag.Message[0:pgpStart]
+	}
+	tag.Signature = &CommitGPGSignature{
+		Signature: ref["contents:signature"],
+		// TODO: don't know what to do about Payload. Is it relevant in
+		// this context?
+		Payload: "",
```

Comment:
note: unsure of (1) what the payload should look like, (2) if it's even needed in this context. To me it _seemed_ like it was pulled out as part of extracting commit metadata and ended up in the tag also "as a bonus". 

It _seems_ to me like the payload _could_ be synthesized from the fields, but unsure if it's needed. Would appreciate guidance.

Reply:
There is a `convertPGPSignature` method maybe helpful.

Reply:
With 083ea22d8 the code produces a payload for annotated tag signatures. I also added a unit test.
I'm still a bit unsure of the _exact_ payload appearance. I did try to copy the `convertPGPSignatureForTag` code to the best of my ability. An extra careful review eye on this would be appreciated.

Reply:
Not sure why the `TestPersistableChannelQueue` test fails in PR. I cannot see that any of my changes would affect that. Could it be a race condition or such?